### ISSUE TITLE
Generic type issue with new useForm hook

### ISSUE THREAD

**Issue by alexturpin** [mantinedev/mantine#1107]
### What package has an issue

@mantine/form

### Describe the bug

When passing a type generic to the new `@mantine/form` `useForm` hook, the types for the `validate` object are not inferred correctly. All `value` types become `unknown`.

[Sandbox link](https://codesandbox.io/s/strange-elgamal-vd0ern?file=/src/App.tsx)

<img width="547" alt="Screen_Shot_2022-03-14_at_19 39 32" src="https://user-images.githubusercontent.com/134339/160267490-85732125-6c36-4af3-8549-786b0dcfd543.png">

Furthermore, the `unknown` types are strictly enforced by TypeScript, e.g. it's not possible to type them in the function parameter because that becomes incompatible with `unknown`.
 
There has already been some conversation around this [in Discord](https://discord.com/channels/854810300876062770/861521000985264138/952894020043014175), but I wanted to document the issue here because in its current state the hook is very hard to use with any types that cannot be inferred from `initialValues`.

### In which browser did the problem occur

N/A

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/strange-elgamal-vd0ern?file=/src/App.tsx

### Do you know how to fix the issue

No

### Are you willing to participate in fixing this issue and create a pull request with the fix

Yes

### Possible fix

I had a look at the types and I'm not sure quite sure why this is happening. I would be happy to help make a PR but I can't quite figure out what the PR would look like for now 😅 if anyone has a hint as to what to try, I will be happy to give it a shot!

**Comment by alexturpin** [comment]
For now, to avoid using the old hook, I am casting manually in every validate function, e.g.:

```
  const { onSubmit, values, getInputProps, validateField } = useForm<PersonalInfoFormValues>({
    initialValues,
    validate: {
      firstName: (v) => {
        const value = v as PersonalInfoFormValues["firstName"]
        return value.trim() === "" ? "This field is required" : null
      },
    },
  })
```

**Comment by jacksonblankenship** [comment]
@alexturpin I know this doesn't answer your question, but I always roll with react-hook-form. This isn't a knock on Mantine's hooks. Personally, I haven't tried them, but rhf is really good at what it does - including inferring types.

**Comment by alexturpin** [comment]
> @alexturpin I know this doesn't answer your question, but I always roll with react-hook-form. This isn't a knock on Mantine's hooks. Personally, I haven't tried them, but rhf is really good at what it does - including inferring types. I'm using it with Mantine on one of my projects right now and it's worked perfectly without any issue at all.

Thanks for the suggestion! I am indeed aware of other form solutions such as RHF but I like Mantine's `useForm` API and I especially like that I can get a simple, low overhead hook for form handling right from Mantine.

**Comment by rtivital** [comment]
To narrow down the issue: we need to create an object with keys that will infer types from provided generic into keys. Seems trivial but I have no idea how to do it – https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFApgZygHgCoBpEDSiKAHqmACZqIDWKAnnMItogLy0NMsB8AFADcAhgBsQ6AFwsAlFIBKKCHABOFTAXyDRU7AG0CAXWnseiAEZw4IlELCmA3gChELxMpRQQypPYC+iIWpbegBuR19HRyUwDAD2ZHQoPntEYEspAHJUDAz8MyFlKQAWACZEX2kwxyEAOjS4apr85SA

**Comment by rtivital** [comment]
Alright, I've managed to fix it in 4.0.9 release. It should work fine now. The solution was to replace Record – https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFApgZygHgCoD4AUAbgIYA2I6AXItgJTUDeiA2gAqIxIDWKAnnMBoBdakTLVsbIbUQBeXIgBGcOKRTEwAX0QMAUIgOIATiiggjSBtuJpEG3gG5dm3bogIMducnRR8TYBVqAHJUDGCAGiViI2oAFgAmRE1aJ11iADpAuHSMxRigA

**Comment by alexturpin** [comment]
> Alright, I've managed to fix it in 4.0.9 release. It should work fine now. The solution was to replace Record 

You're the absolute best Vitaly, thank you!
